### PR TITLE
Ckan media integration

### PR DESCRIPTION
**Pull request by jeremy-doghouse**
Hey @srowlands & @gollyg 

This PR address all but one issue currently in the queue, @jorgegc-doghouse and I are also looking into a fix for the field install via hook_enable but that is really just a best practice thing.

Issues that can potentially be closed:
#1 - Validator added to enforce HTTPS if API key present when adding the base endpoint url.
#3 - Pluggable visualisations - Custom charts are found in the display module while the media module contains the default visualisation of a simple table.
#4 - Visualisation selector available which ajax loads visualisation config into the field widget
#5 - All visualisation configuration is serialised and saved against the custom field type
#6 - Custom configuration per visualisation with a helper to load in generic form elements for key selection
#7 - All visualisations output is parsed and displayed as a table by default, JS adds charts ontop of that
#8 - Visualisation settings attached to the table via data attributes, this is implemented by the visualisation plugin view
#9 - Bar chart visualisation complete and includes all required fields
#10 - Spline chart visualisation complete and includes all required fields
#11 - Line chart visualisation complete and includes all required fields

Bonus extras:
* Added some field formatting settings for settings that should be field wide (eg available to all visualisations), this flexible with what we use it for but to start of with is: including export dimensions to assist with ENV-257, ability to define a palette for the visualisation and limit the number of ticks on an X or Y axis (testing with different datasets revealed this as a requirement). These formatter settings get merged with visualisation settings and get passed to the visualisation view handler.
* Various fixes and improvements to chart JS
* Ensured the link to a ckan file type is the public url to the dataset, solving ENV-140
* Improvements to the parsing class including adding attributes and the option to output as a renderable array
* If unable to fetch results from api, error message is rendered in place of the visualisation, the text for this can be overridden with a variable.
* Much much more...

*The module currently works great!* Simply enable govcms_ckan_media and add a url to a public dataset and follow the wizard.  

A bit more polish is required here and there but a solid framework is in place. All feedback welcome.

Side notes: 
* Key selection have defaulted to none selected, this really feels better than selecting all, potentially containing keys that will break the output.
* Vertical/Horizontal is select rather than radios, it just works better and consumes less screen real estate.

Enjoy :)

### CODE REVIEW

**Comment by srowlands** [comment]
Brilliant. Thanks @jeremy-doghouse, will dig in. :+1:

File: govcms_ckan.admin.inc [hunk 1]
```diff
@@ -45,6 +45,14 @@ function govcms_ckan_settings_form() {
  * @see govcms_ckan_settings_form()
  */
 function govcms_ckan_settings_form_validate($form, &$form_state) {
+  // If an API key is in use, enforce https.
+  if (!empty($form_state['values']['govcms_ckan_api_key'])) {
+    $url_parts = parse_url($form_state['values']['govcms_ckan_endpoint_url']);
```
Comment (gollyg) [inline govcms_ckan.admin.inc]: What about using https://api.drupal.org/api/drupal/includes%21file.inc/function/file_uri_scheme/7 here

File: govcms_ckan.module [hunk 2]
```diff
@@ -158,6 +164,35 @@ function govcms_ckan_get_plugin_config($name, $form, $form_state, $config) {
 }
 
 /**
+ * Helper function to return a specific visualisation plugin config form.
+ *
+ * @param string $name
+ *   The name of the visualisation plugin to return.
+ * @param array $file
```
Comment (gollyg) [inline govcms_ckan.module]: File or object?

Comment (jeremy-doghouse) [inline govcms_ckan.module]: Object :) 
Also the description for that comment was wrong, returns view not form. Both fixed in #14

File: modules/govcms_ckan_media/plugins/visualisation/table.inc [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * Returns a renderable array that represents the block content.
+ */
+function govcms_ckan_media_table_view($file, $display, $config) {
+  $element = array();
+  $chart_class = 'ckan-table';
+  $parser = govcms_ckan_dataset_parser();
+  $response = govcms_ckan_client_request_records($file->resource_id);
+
+  // If failure, provide error message.
+  if ($response->valid !== FALSE) {
```
Comment (gollyg) [inline modules/govcms_ckan_media/plugins/visualisation/table.inc]: Not sure if this is relevant, but http://docs.ckan.org/en/latest/api/#making-an-api-request talks about valid responses against ckan. The parser currently checks the http status code, but the docs seem to indicate it is better to check against the success key in the response. Doesn't form part of this PR anyway, but just noting.

**Pull request merged**